ipam: Wait for ENI netlink interface before configuring routes

pippolo84 opened the pull request:
In ENI IPAM mode, the ENIs are created by the operator. At the same
time, on each node, the agents configure the v4 and v6 rules and routes
for the Cilium router, after retrieving its IPs from either k8s or the
filesystem. In order to do so, each agent query netlink to get the
ifindex of the interface with the router IP, given its MAC address.
Unfortunately this behavior is racy, since the agent might query netlink
too soon, when the ENI is not yet up and running. This leads to the
following error from netlink:

"daemon creation failed: failed to configure router IP rules and routes:
unable to find ifindex for interface MAC: interface with MAC ... not
found"

that ultimately stops the daemon startup.

To address this, let's poll netlink and wait for the ENI netlink
interface to show up before going ahead with routes and rules
configuration.

Fixes: #37948

~Depends on https://github.com/cilium/cilium/pull/41783~

pippolo84 committed:
ipam: Rely on the daemon context in datapath IPs allocation

Now that the waitForENI function accepts a context, pass the daemon
context as an argument. This way the startup procedure remains
responsive to a context cancellation.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

pippolo84 committed:
aws/eni: Improve metadata client to accept a context

Accepting a context from the caller instead of hard-coding a
context.TODO() allows for early termination of the IMDS queries when
needed.

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>

gandro commented on daemon/cmd/ipam.go:
```diff
@@ -146,6 +149,28 @@ func reallocateDatapathIPs(logger *slog.Logger, alloc ipamAllocateIP, fromK8s, f
+	}
+
+	findENIByMAC := func(ctx context.Context) (bool, error) {
+		macs, err := metadata.GetMACs(ctx)
```
So the commit message says that netlink might not have discovered the interface yet. But here we are querying the metdata service, which doesn't guarantee that the local kernel actually has attached the interface (i.e. loading the needed driver for the interface can take a few milliseconds, even if the hypervisor has attached it). 

While this makes the race far less likely, I do wonder if we should query netlink here instead and ensure that the expected mac address shows up in `netlink.LinkList()`.

liyihuang replied:
Just put my 2 cents here, I found this implementation in AWS CNI. That's why I suggested this way. I guess the race conditions happened from AWS operations than kernel. I guess it's because ENI attachement happens async from operator and got the MAC address. Then agent picks it up from CR.

gandro replied:
Right, thanks for the context. I'm also saying this because for non-router ENIs we actually check the netlink interfaces, not the metadata service before we try to allocate from it:

https://github.com/cilium/cilium/blob/de893e6330cc5b0d4b8fbc3d839a6b54a96c0733/pkg/ipam/crd_eni.go#L75-L92

gandro replied:
Maybe we could also execute the above code not in the background, but always before we hand out IPs for a newly attached ENI. That way we wouldn't have to change `allocateDatapathIPs` and instead have the IPAM subsystem guarantee that whenever a we allocate an IP that the corresponding host interface is attached and ready.

pippolo84 replied:
Discussed this offline with @gandro . We decided to keep it simple for now and change `allocateDatapathIPs` to wait for the netlink interface to show up. I've updated the PR but kept the first commit since it was an useful improvement to the aws metadata client (though it is not needed for the fix anymore).

pippolo84 replied:
I'll also open an issue to keep track of the approach suggested by @gandro since it is a nicer solution in the long term.

gandro reviewed (commented):
Thanks for the fix! One question/comment w.r.t. the approach

pippolo84 committed:
ipam: Wait for ENI netlink iface before configuring routes

In ENI IPAM mode, the ENIs are created by the operator. At the same
time, on each node, the agents configure the v4 and v6 rules and routes
for the Cilium router, after retrieving its IPs from either k8s or the
filesystem. In order to do so, each agent query netlink to get the
ifindex of the interface with the router IP, given its MAC address.
Unfortunately this behavior is racy, since the agent might query netlink
too soon, when the ENI is not yet up and running. This leads to the
following error from netlink:

"daemon creation failed: failed to configure router IP rules and routes:
unable to find ifindex for interface MAC: interface with MAC ... not
found"

that ultimately stops the daemon startup.

To address this, let's poll netlink and wait for the ENI netlink
interface to show up before going ahead with routes and rules
configuration.

Fixes: #37948

Signed-off-by: Fabio Falzoi <fabio.falzoi@isovalent.com>
